### PR TITLE
[go_router] Allow any number of the same page on the stack

### PR DESCRIPTION
**Pull request by koji-1009**
Restore the function to push the same page multiple times, which was supported in v3.1.1.

Fixes https://github.com/flutter/flutter/issues/107045

Refs
* https://github.com/flutter/packages/blob/go_router-v3.1.1/packages/go_router/lib/src/go_router_delegate.dart#L264
* https://github.com/csells/go_router/commit/800d29ae9ddc6c26573248f79042d3826f5175fd#diff-a894fc0967a6e4235d9deb900b9454f67082e9836146c1024f863bf73f578cb7R234

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by ValentinVignal** [comment]
I think it is a duplicate of https://github.com/flutter/flutter/issues/107045

**Review by chunhtai — COMMENTED** [review]
I started to think support the push in the declarative way may be the wrong decision. as soon as a page is pushed to the stack, the stack will not be in sync with the url any more. 
In this particular use case where you want to push the same page, does it make more sense to be a pageless route? that is, pushing use the Navigator.of(context). The difference is the pageless route will belong to the top most page route and will be move/remove with that page route

**Comment by koji-1009** [comment]
On GitHub, after opening the top of the repository, you can open the list of pull requests and open the pull request details.
Also, using this PR as an example, clicking on the `flutter:main` branch link will open the top of the repository. A use case would be to implement this process in Flutter Web/Android/iOS.
At this point, we have the choice between `context.go` and `context.push`. The problem in this case is not `go` but `push`, so we want to back to the last open page.

---

If we don't want the delegate to have a Map, why not change the process so that the `pageKey` has a random suffix (e.g. UUID). Having `state.pageKey` as the Route key is the preferred method for most applications that use go_router.

File: packages/go_router/test/delegate_test.dart [hunk 1]
```diff
@@ -56,6 +57,34 @@ void main() {
...
+      'It should return different pageKey when push is called',
+      (WidgetTester tester) async {
+        final GoRouter goRouter = await createGoRouter(tester);
+        expect(goRouter.routerDelegate.matches.matches.length, 1);
+        expect(
+          goRouter.routerDelegate.matches.matches[0].pageKey,
+          null,
+        );
+
+        goRouter.push('/a');
```
Comment (chunhtai) [inline packages/go_router/test/delegate_test.dart:71]: can you do an `await tester.pumpAndSettle()` in between every push to verify it does not crash?

File: packages/go_router/lib/src/delegate.dart [hunk 2]
```diff
@@ -45,10 +45,27 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
 
   RouteMatchList _matches = RouteMatchList.empty();
+  final Map<String, int> _pushCounts = <String, int>{};
 
   /// Push the given location onto the page stack
   void push(RouteMatch match) {
-    _matches.push(match);
+    // remap the pageKey so allow any number of the same page on the stack
```
Comment (chunhtai) [inline packages/go_router/lib/src/delegate.dart:52]:
```suggestion
    // Remap the pageKey to allow any number of the same page on the stack
```

File: packages/go_router/lib/src/delegate.dart [hunk 3]
```diff
@@ -45,10 +45,27 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
 
   RouteMatchList _matches = RouteMatchList.empty();
+  final Map<String, int> _pushCounts = <String, int>{};
 
   /// Push the given location onto the page stack
```
Comment (chunhtai) [inline packages/go_router/lib/src/delegate.dart:50]:
```suggestion
  /// Pushes the given location onto the page stack
```

**Review by chunhtai — APPROVED** [review]
LGTM, just some nitpicks

**Review by johnpryan — APPROVED** [review]

**Review by prince-kumar — COMMENTED** [review]
LGTM

**Comment by chunhtai** [comment]
not sure why the check failed, can you rebase and try again?

File: packages/go_router/test/delegate_test.dart (hunk 1 above)
Comment (koji-1009) [reply packages/go_router/test/delegate_test.dart:71]: Thanks for the review, I fixed it in 2ee5fd6.

**Comment by koji-1009** [comment]
https://github.com/flutter/plugins/pull/6146

**Comment by stuartmorgan** [comment]
This change must *not* be landed until the underlying problem with 4.2.1 is fixed, or it will re-break everyone that was broken by 4.2.1 (since it'll publish a new version with the same bug).

@johnpryan @chunhtai please don't approve any more go_router PRs until the fix for the breakage is fixed, to avoid any accidents.

**Review by stuartmorgan — DISMISSED** [review]
Adding a request changes until the above has been resolved to ensure this doesn't accidentally land.

**Review by johnpryan — APPROVED** [review]

**Pull request merged**